Fix #438, add three PDI deactivation option through CMake (#543)
pdidev/pdi: pdi/docs/Using_PDI.md

For C make sure that source files that use %PDI API are including `pdi.h` header file.
For Fortran make sure that source files that use %PDI API are using `%PDI` module file (`USE %PDI`).

> %PDI can be disabled by using the `no-pdi` directory instead of the `pdi` directory.

Author: Less precise. Clarify that no elements of the pdi/pdi folder are used, when using no-pdi.

> %PDI can be disabled by using the `no-pdi` directory instead of the `pdi` directory.

Reviewer: Users do not necessarily see a "PDI directory". If they use a PDI installed by apt, there is no such PDI directory.

> You can use the `find_package` method, which does not require to modify the 

Reviewer: I would point to the `add_subdirectory` strategy first, and only document this as an alternative.

Author: The `find_package` method is the one requiring less code change from the user.

Reviewer: Indeed. That's why I would advise to use the `add_subdirectory` strategy as the recommended one when possible, since it is the least fragile. And then, you can say "if modifying the CMakeList.txt is not possible" and point to the `find_package` as a fallback.

target `CMakeLists.txt` but requires to compile with an added argument pointing to the no-pdi folder. 

Use a `CMakeLists.txt` similar to `no-pdi_exampleTargetCMakeLists_findpackage.txt` for your target, 
then use the following:
```bash

> cmake . -DCMAKE_MODULE_PATH="/<full>/<path>/<to>/pdi/no-pdi/cmake"

Reviewer: ```suggestion
cmake . -PDI_ROOT="/<full>/<path>/<to>/pdi/no-pdi/cmake"
```

When switching from FindPDI to PDIConfig

Author: This method implies modifying the user target CMakeLists, using :
`include(${PDI_ROOT}/PDIConfig.cmake)`
instead of :
`find_package(PDI 1.0.0 REQUIRED COMPONENTS C)`

```

> Alternatively, you can use the `include`/`target_include_directories` method, 

Reviewer: ```suggestion
Alternatively, you can use the `add_subdirectory` method, 
```

Author: The example file does not have the text `add_subdirectory` but `include`.

which does not require to pass an additional argument at compilation 

> but requires to add ad option to the target `CMakeLists.txt`, 

Reviewer: ```suggestion
but requires to slightly modify your `CMakeLists.txt`, 
```

Author: Less precise. Needs to indicate the part of file which is concerned by the change.

to compile with an added argument pointing to the no-pdi folder.

Use a `CMakeLists.txt` similar to `tests/no-pdi/CMakeLists.txt` with `PDI_ROOT` for your target, 
then use the following:
```bash
cmake . -PDI_ROOT="/<full>/<path>/<to>/pdi/no-pdi/cmake"
```

> 

Reviewer 2: I think `no-pdi` merits a separate section. Maybe at the end of the doc.

> In this case, make sure to modify your target application `CMakeLists.txt` following 
> the file `no-pdi_exampleTargetCMakeLists_usingsubdir.txt` included in the example directory,

Reviewer: ```suggestion
In this case, make sure to modify your target application `CMakeLists.txt` as illustrated in 
`example/CMakeLists.txt`,
```

2. `plugin_path` subtree in specification tree: \ref plugin_path_map_node,
3. Relative path of used %PDI shared object `libpdi.so`,
4. `LD_LIBRARY_PATH` environment variable that is colon separated list.

## How to deactivate PDI {#deactivate_pdi}

### Using the no-pdi directory with an available Paraconf

You may copy the no-pdi folder from %PDI in your application repository, and
must add an option (`CMAKE_PREFIX_PATH` in the example below) to your
cmake command to use this mock folder instead of the real PDI.
Use your target `CMakeLists.txt` with `SRCDIR` as the directory containing no-pdi 
(it is the root %PDI directory if you did not move the no-pdi directory) :
```bash
cmake . -DCMAKE_PREFIX_PATH="${CMAKE_PREFIX_PATH:+${CMAKE_PREFIX_PATH}:}${SRCDIR}/no-pdi"
```
%PDI path as `SRCDIR` containing no-pdi in the example above.

If you add this option, Paraconf will only be used if it is found.

> In the case if is not found, only the basic `.yaml` file reader of Paraconf 

Reviewer 2: ```suggestion
In the case it is not found, only the basic `.yaml` file reader of Paraconf 
```
what is the basic yaml reader of paraconf?

> will be enabled to fit with no-pdi, through `PDI_init`.

Reviewer 2: My suggestion:

You may use the no-pdi mock folder instead of the real PDI to compile your application. The only modification you need to make is to add the `CMAKE_PREFIX_PATH` option to your CMake command.
For example:
```bash
cmake . -your-usual-cmake-compile-options -DCMAKE_PREFIX_PATH="${CMAKE_PREFIX_PATH:+${CMAKE_PREFIX_PATH}:}<path-to-pdi>/no-pdi"
```
Regarding the non-availability of Paraconf on your system, you need to comment out in your application `#include <paraconf.h>` (in C/C++ as an example). You will also need to comment out the call to `PDI_init`. Eventually, you will need to comment out any calls to the Paraconf API.

2. `plugin_path` subtree in specification tree: \ref plugin_path_map_node,
3. Relative path of used %PDI shared object `libpdi.so`,
4. `LD_LIBRARY_PATH` environment variable that is colon separated list.

## How to deactivate PDI {#deactivate_pdi}

You can use the `PDIConfig` method, 
which does not require to pass an additional argument at compilation 
but requires to add an option to the target `CMakeLists.txt`, 
which must be enabled through an option at compilation.

> 

Reviewer 3: I suggest to add a subsection for each different method or number list of the different method

2. `plugin_path` subtree in specification tree: \ref plugin_path_map_node,
3. Relative path of used %PDI shared object `libpdi.so`,
4. `LD_LIBRARY_PATH` environment variable that is colon separated list.

## How to deactivate PDI {#deactivate_pdi}

> You can use the `PDIConfig` method, 

Reviewer 2: if this is one of the three options, it should be inside a dedicated paragraph like the other two options.

Author: Now that both options use the "config" file (and not he "find" file anymore), this name represents both of the currently included alternatives.
Clarified in the next push.

which does not require to pass an additional argument at compilation 
but requires to add an option to the target `CMakeLists.txt`, 
which must be enabled through an option at compilation.

### Using a no-pdi included in the currently used pdi repository

Use a `CMakeLists.txt` similar to `example/CMakeLists.txt` with `EXAMPLES_WITHOUT_PDI` for your target, 
then use the following:
```bash
cmake . -DEXAMPLES_WITHOUT_PDI=ON
```

### Using a specific no-pdi through a full path

> Alternatively, you can also use the `PDIConfig` method 

Author: Now that both options use the "config" file (and not he "find" file anymore), this name represents both of the currently included alternatives.
Clarified in the next push.
